Default position_offset_gap to 100

nik9000 opened the pull request:
This is much more fiddly than it looks because of the way position_offset_gap
is applied in StringFieldMapper. Instead of setting the default to 10 its
simpler to make sure that all the analyzers default to 10 and that
StringFieldMapper doesn't override the default unless the user specifies
something different.

Closes #7268

nik9000 commented:
WIP because it needs tests and more testing. This is way way harder then it ought to be.

nik9000 commented:
Oh and it needs documentation changes and breaking changes documentation.

rmuir commented:
Which one is it? the position gap or the offset gap? Lucene has both, and they both have different meanings. Changing the former to this value makes sense, the latter will break many things.

rmuir commented:
And please, please, we have to rename this, because its totally meaningless. positions and offsets are separate things...

nik9000 commented:
I understand the rename. It's the positions.
On Jul 29, 2015 8:02 PM, "Robert Muir" <notifications@github.com> wrote:

> And please, please, we have to rename this, because its totally
> meaningless. positions and offsets are separate things...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/12544#issuecomment-126132622>
> .
>

nik9000 commented:
Does anyone have an opinion on what would be ok behavior from a backwards compatibility standpoint?

nik9000 commented:
> Does anyone have an opinion on what would be ok behavior from a backwards compatibility standpoint?

Talked with @dakrone and we decided that the most correct thing as to make ```position_offset_gap``` be immutable once the index is created even if it isn't set. So indexes created before 2.0.0-beta1 will always have a default ```position_offset_gap``` of ```0``` regardless of which version of elasticsearch operates on them. Indexes created by versions of Elasticsearch on or after 2.0.0-beta1 will default to ```10``` like the issue says they should.

nik9000 commented:
I believe this is ready for review. @jpountz - its more fiddly than I expected at first so it might be worth a review from you or someone who's pretty experienced with mapping.

@mute - I _think_ this is the right way to do #12538. Its much more complicated than it looked and certainly wasn't really low hanging fruit.

nik9000 commented:
I'd love this in 2.0.0 at some point. It'll need a review soon if it is going to.

mikemccand commented on docs/reference/mapping/types/core-types.asciidoc:
```diff
@@ -150,7 +150,7 @@ limit to `32766 / 3 = 10922` since UTF-8 characters may occupy at most 3
 
 |`position_offset_gap` |Position increment gap between field instances
-with the same field name. Defaults to 0.
+with the same field name. Defaults to the default of the analyzer which is 10.
```
10 seems so arbitrary, but I'm OK with it.  I assume you didn't pick 1 because then a sloppy phrase query with slop=1 would incorrectly match...

But, can we explain why it's non-zero here?  E.g. just add "... so that proximity queries like PhraseQuery do not incorrectly match across 2 values of a multi-valued field"?

nik9000 replied:
Yeah! I'll do that.

10 is totally arbitrary. Its much more than 1 which is too low. I'd be happy to change it to 1024 or some other magic number. I just want it much much more than 1.

mikemccand commented on docs/reference/analysis/analyzers/custom-analyzer.asciidoc:
```diff
@@ -20,8 +20,8 @@ filters.
-|`position_offset_gap` |An optional number of positions to increment 
-between each field value of a field using this analyzer.
+|`position_offset_gap` |An optional number of positions to increment
+between each field value of a field using this analyzer. Defaults to 10.
```
Also explain the motivation for 10 here?

nik9000 replied:
Sure.

mikemccand commented on core/src/main/java/org/elasticsearch/index/mapper/core/StringFieldMapper.java:
```diff
@@ -102,7 +120,7 @@ public Builder ignoreAbove(int ignoreAbove) {
         @Override
         public StringFieldMapper build(BuilderContext context) {
-            if (positionOffsetGap > 0) {
+            if (positionOffsetGap != -1) {
```
Change -1 to POSITION_OFFSET_GAP_USE_ANALYZER?

nik9000 replied:
Oh! Yeah.

clintongormley commented:
I agree that 10 is small, and could quite easily overlap with a typical slop value.  Personally i'd go for eg 100

mikemccand commented on core/src/main/java/org/elasticsearch/index/mapper/core/StringFieldMapper.java:
```diff
@@ -155,6 +173,9 @@ public StringFieldMapper build(BuilderContext context) {
                 } else if (propName.equals("position_offset_gap")) {
                     builder.positionOffsetGap(XContentMapValues.nodeIntegerValue(propNode, -1));
+                    if (builder.positionOffsetGap < 0) {
+                        throw new MapperParsingException("positions_offset_gap less than 0 aren't allowed.");
```
I know the exc we throw here effectively discards `builder` but I think it's less smelly to first assign to a local "untrusted" var first, e.g. `int newPositionOffsetGap`, and check whether that's < 0, before calling `builder.positionOffsetGap(newPositionOffsetGap)`.

nik9000 replied:
Yeah - I thought about that. I'll make that change.

mikemccand commented on docs/reference/migration/migrate_2_0.asciidoc:
```diff
@@ -194,6 +194,7 @@ A `RoutingMissingException` is now thrown instead.
 * The delete mappings API has been removed. Mapping types can no longer be deleted.
 * The `ignore_conflicts` option of the put mappings API has been removed. Conflicts can't be ignored anymore.
 * The `binary` field does not support the `compress` and `compress_threshold` options anymore.
+* The default `position_offset_grap` is now 10. Indexes created in Elasticsearch 2.0.0 will default to using 10 and indexes created before that will continue to use the old default of 0.
```
I think also explain the motivation for the change here?

nik9000 commented:
Ok - I'll make these changes. I don't think I'll have time today because I'm in class and can't really concentrate. But tonight or tomorrow "morning".

mikemccand commented on core/src/main/java/org/elasticsearch/index/analysis/AnalysisService.java:
```diff
@@ -215,19 +216,25 @@ public AnalysisService(Index index, @IndexSettings Settings indexSettings, @Null
             if (analyzerF instanceof NamedAnalyzer) {
+                // if we got a named analyzer back, use it...
                 analyzer = (NamedAnalyzer) analyzerF;
+                if (overridePositionOffsetGap >= 0 && analyzer.getPositionIncrementGap(analyzer.name()) != overridePositionOffsetGap) {
```
Maybe add some comments here explaining?  I'm confused...

nik9000 replied:
Yeah - this was confusing when I wrote it. Its funky and deserves more comments....

nik9000 replied:
Ok - added some comments above.

mikemccand commented:
Thanks @nik9000 this is a great change (so prox aware queries never match across 2 values of a multi-valued field), I just left some minor comments.

Can you open a follow-on issue to rename `positionOffsetGap`?  Need not block this good change, but this is extremely confusing :)  As @rmuir said, position gap and offset gap are wildly different things!

nik9000 commented:
> Can you open a follow-on issue to rename positionOffsetGap? Need not block this good change, but this is extremely confusing :) As @rmuir said, position gap and offset gap are wildly different things!

#12562

nik9000 commented:
Swapped 2.0 for 2.1 because we're too late in the 2.0 release cycle to get this merged there.

I'm going to pick this one back up in a few minutes and have another read through. I'll see if I can address that last open comment of @mikemccand and rebase. And I'll change all the 2.0s into 2.1s in the code.

nik9000 commented on docs/reference/mapping/types/core-types.asciidoc:
```diff
@@ -0,0 +1,651 @@
+[[mapping-core-types]]
```
grrr - how did I add a whole new file....

nik9000 replied:
Ah. I figured it out - rebase woes. Fixing.

nik9000 commented:
@mikemccand  - would you mind having another look at this? Its now ready for review again based on 2.1.

mikemccand commented:
LGTM, thanks @nik9000!

s1monw commented:
@nik9000 this looks good I think this should go into 2.0 though

nik9000 commented:
> @nik9000 this looks good I think this should go into 2.0 though

Ok - I'll merge to 2.1 now and backport it. I'm not super comfortable sticking it in 2.0 but I'll try because you want it in there.

nik9000 commented:
I merged this into master but github doesn't recognize it - probably because I merged it locally and pushed. That failed so I rebased onto elastic's master and squashed. At this point github's must have lost any connection to the patch.

I'll leave this open to work the backport to 2.0. I'll close it once I've merged there.

xuzha commented on core/src/main/java/org/elasticsearch/index/mapper/core/StringFieldMapper.java:
```diff
@@ -62,15 +64,36 @@
+         * values.
+         */
+        public static final int POSITION_OFFSET_GAP = 100;
+        public static final int POSITION_OFFSET_GAP_PRE_2_1 = 0;
```
This change is going to 2.0 branch right? Do we need to change this to 2_0?

nik9000 replied:
Sadly, yup. I had originally targeted 2.0, then gave up on making it in, then @s1monw said we should jam it in there anyway. So I got what I had reviewed merged as quickly as I could. It's looking like it'll be reasonably quick to merge to 2.0 so I'll do that and _then_ go fix master so it makes sense.

xuzha commented on docs/reference/migration/migrate_2_1.asciidoc:
```diff
@@ -25,3 +25,13 @@ GET /my_index/_search?scroll=2m
+=== Mapping changes
+
+==== position_offset_gap
+The default `position_offset_grap` is now 100. Indexes created in Elasticsearch
```
position_offset_grap -> position_offset_gap

nik9000 replied:
Ok - Will fix this one right quick.

nik9000 replied:
840b4cf0e110deba6f1222539789209ec906f05a

nik9000 commented:
And merged to 2.0. For my last trick on this pull request I'll fix master so that it has the same version range checks as 2.0. And then finally, finally, this is done.

nik9000 commented:
> And merged to 2.0. For my last trick on this pull request I'll fix master so that it has the same version range checks as 2.0. And then finally, finally, this is done.

Scratch that, I'll send that as another pull request.
